messaging/src/main/java/org/axonframework/commandhandling/gateway/AbstractRetryScheduler.java

```diff
@@ -0,0 +1,175 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static org.axonframework.common.BuilderUtils.assertNonNull;
+import static org.axonframework.common.BuilderUtils.assertStrictPositive;
+
+/**
+ * An abstract base class for {@link RetryScheduler}s. This class provides methods to do the actual rescheduling and
+ * decide if a given {@link Throwable} is explicitly transient.
+ */
+public abstract class AbstractRetryScheduler implements RetryScheduler {
+
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private static final int DEFAULT_MAX_RETRIES = 1;
+
+    private final ScheduledExecutorService retryExecutor;
+    private final int maxRetryCount;
+
+    protected AbstractRetryScheduler(Builder builder) {
+        this.retryExecutor = builder.retryExecutor;
+        this.maxRetryCount = builder.maxRetryCount;
+    }
+
+    /**
+     * Indicates whether the given {@code failure} is clearly non-transient. That means, whether the
+     * {@code failure} explicitly states that a retry of the same Command would result in the same failure to
+     * occur
+     * again.
+     *
+     * @param failure the exception that occurred while processing a command
+     * @return {@code true} if the exception is clearly non-transient and the command should <em>not</em> be
+     * retried, or {@code false} when the command has a chance of succeeding if it retried.
+     */
+    protected boolean isExplicitlyNonTransient(Throwable failure) {
+        return failure instanceof AxonNonTransientException
+                || (failure.getCause() != null && isExplicitlyNonTransient(failure.getCause()));
+    }
+
+    /**
+     * Schedule the provided task to run after the given interval.
+     *
+     * @param commandDispatch the {@link Runnable} to schedule.
+     * @param interval        the number of milliseconds delay.
+     * @return {@code true} if the task was accepted for scheduling, {@code false} otherwise.
+     */
+    protected boolean scheduleRetry(Runnable commandDispatch, long interval) {
+        try {
+            retryExecutor.schedule(commandDispatch, interval, TimeUnit.MILLISECONDS);
+            return true;
+        } catch (RejectedExecutionException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Compute the amount of milliseconds delay until the next retry, given the information passed.
+     *
+     * @param commandMessage the command that was sent (and failed).
+     * @param lastFailure    the last failure that caused this retry scheduler to be called.
+     * @param failures       a {@link List} of all failures up to now.
+     * @return the number of milliseconds to wait until retrying.
+     */
+    protected abstract long computeRetryInterval(CommandMessage commandMessage,
+                                                 RuntimeException lastFailure,
```

Comment:
Shouldn't the type of the second parameter be `Throwable` (or `Exception`) instead of `RuntimeException`? :thinking:

Reply:
See previous comment on same issue.

Reply:
Allard used his overrule powers (with fair reasoning), so I'll resolve this comment for you.